opentelemetry-api/src/opentelemetry/context/propagation/tracecontexthttptextformat.py

```diff
@@ -127,25 +124,42 @@ def inject(
             )
 
 
-def _parse_tracestate(string: str) -> trace.TraceState:
-    """Parse a w3c tracestate header into a TraceState.
+def _parse_tracestate(header_list: typing.List[str]) -> trace.TraceState:
+    """Parse one or more w3c tracestate header into a TraceState.
 
     Args:
         string: the value of the tracestate header.
 
     Returns:
         A valid TraceState that contains values extracted from
         the tracestate header.
+
+        If the format of one headers is illegal, all values will
+        be discarded and an empty tracestate will be returned.
+
+        If the number of keys is beyond the maximum, all values
+        will be discarded and an empty tracestate will be returned.
     """
     tracestate = trace.TraceState()
-    for member in re.split(_DELIMITER_FORMAT_RE, string):
-        match = _MEMBER_FORMAT_RE.match(member)
-        if not match:
-            raise ValueError("illegal key-value format %r" % (member))
-        key, _eq, value = match.groups()
-        # typing.Dict's update is not recognized by pylint:
-        # https://github.com/PyCQA/pylint/issues/2420
-        tracestate[key] = value  # pylint:disable=E1137
+    for header in header_list:
+        for member in re.split(_DELIMITER_FORMAT_RE, header):
+            # empty members are valid, but no need to process further.
+            if not member:
+                continue
+            match = _MEMBER_FORMAT_RE.fullmatch(member)
+            if not match:
+                # TODO: log this?
+                return trace.TraceState()
+            key, _eq, value = match.groups()
+            if key in tracestate:  # pylint:disable=E1135
+                # duplicate keys are not legal in
+                # the header, so we will remove
+                return trace.TraceState()
+            # typing.Dict's update is not recognized by pylint:
+            # https://github.com/PyCQA/pylint/issues/2420
+            tracestate[key] = value  # pylint:disable=E1137
+    if len(tracestate) > _TRACECONTEXT_MAXIMUM_TRACESTATE_KEYS:
```

Comment:
probably DoS could be achieved other ways, like adding a huge list of values that need to be parsed by regex. but agreed this could be a quick shortcut out.